// @triplit/db/dist/hybrid-clock.js
// TODO use KV / make durable
export class HybridLogicalClock {
    lastPhysicalTimeMs;
    lastLogicalTime;
    clientId = '';
    constructor(options) {
        this.clientId = options.clientId;
        this.lastPhysicalTimeMs = options.lastPhysicalTimeMs || 0;
        this.lastLogicalTime = options.lastLogicalTime || 0;
    }
    get currentPhysicalTimeMs() {
        return Date.now();
    }
    current() {
        return [this.currentPhysicalTimeMs, this.lastLogicalTime, this.clientId];
    }
    next() {
        const physicalTimeMs = this.currentPhysicalTimeMs;
        if (physicalTimeMs > this.lastPhysicalTimeMs) {
            this.lastPhysicalTimeMs = physicalTimeMs;
            this.lastLogicalTime = 0;
        }
        else {
            this.lastLogicalTime += 1;
        }
        return [this.lastPhysicalTimeMs, this.lastLogicalTime, this.clientId];
    }
    updatePhysicalTimeMs(physicalTimeMs) {
        this.lastPhysicalTimeMs = physicalTimeMs;
    }
    static compare(a, b) {
        if (a[0] === b[0]) {
            if (a[1] === b[1]) {
                return a[2].localeCompare(b[2]);
            }
            return a[1] - b[1];
        }
        return a[0] - b[0];
    }
    static MIN = Object.freeze([0, 0, '']);
}
