// @triplit/db/dist/permissions.js
export function getCollectionPermissions(schema, collectionName) {
    if (!schema)
        return undefined;
    const collection = schema[collectionName];
    if (!collection)
        return undefined;
    return collection.permissions;
}
export function isReadPermissionOperation(operation) {
    return operation === 'read';
}
export function isWritePermissionOperation(operation) {
    return !isReadPermissionOperation(operation);
}
export function permissionsEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (const key of keysA) {
        const permissionsA = a[key];
        const permissionsB = b[key];
        if (!operationPermissionsEqual(permissionsA, permissionsB))
            return false;
    }
    return true;
}
function operationPermissionsEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    if (!permissionDefinitionEqual(a.read, b.read))
        return false;
    if (!permissionDefinitionEqual(a.insert, b.insert))
        return false;
    if (!permissionDefinitionEqual(a.update, b.update))
        return false;
    if (!permissionDefinitionEqual(a.postUpdate, b.postUpdate))
        return false;
    if (!permissionDefinitionEqual(a.delete, b.delete))
        return false;
    return true;
}
function permissionDefinitionEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    if (!permissionFilterEqual(a.filter, b.filter))
        return false;
    return true;
}
function permissionFilterEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    // TODO: this can be more precise
    return JSON.stringify(a) === JSON.stringify(b);
}
export function rolesEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (const key of keysA) {
        const roleA = a[key];
        const roleB = b[key];
        if (!roleEqual(roleA, roleB))
            return false;
    }
    return true;
}
export function roleEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    if (!permissionMatcherEqual(a.match, b.match))
        return false;
    return true;
}
function permissionMatcherEqual(a, b) {
    if (!a && !b)
        return true;
    if (!a || !b)
        return false;
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (const key of keysA) {
        const matcherA = a[key];
        const matcherB = b[key];
        if (JSON.stringify(matcherA) !== JSON.stringify(matcherB))
            return false;
    }
    return true;
}
