// @instantdb/react/dist/commonjs/next-ssr/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstantRouteHandler = exports.InstantSuspenseProvider = exports.InstantNextDatabase = exports.getUserFromInstantCookie = exports.getUnverifiedUserFromInstantCookie = void 0;
exports.init = init;
const version_ts_1 = __importDefault(require("../version.js"));
const InstantNextDatabase_tsx_1 = require("./InstantNextDatabase.js");
var getUserFromInstantCookie_ts_1 = require("./getUserFromInstantCookie.js");
Object.defineProperty(exports, "getUnverifiedUserFromInstantCookie", { enumerable: true, get: function () { return getUserFromInstantCookie_ts_1.getUnverifiedUserFromInstantCookie; } });
Object.defineProperty(exports, "getUserFromInstantCookie", { enumerable: true, get: function () { return getUserFromInstantCookie_ts_1.getUserFromInstantCookie; } });
var InstantNextDatabase_tsx_2 = require("./InstantNextDatabase.js");
Object.defineProperty(exports, "InstantNextDatabase", { enumerable: true, get: function () { return InstantNextDatabase_tsx_2.InstantNextDatabase; } });
var InstantSuspenseProvider_tsx_1 = require("./InstantSuspenseProvider.js");
Object.defineProperty(exports, "InstantSuspenseProvider", { enumerable: true, get: function () { return InstantSuspenseProvider_tsx_1.InstantSuspenseProvider; } });
var core_1 = require("@instantdb/core");
Object.defineProperty(exports, "createInstantRouteHandler", { enumerable: true, get: function () { return core_1.createInstantRouteHandler; } });
/**
 *
 * The first step: init your application!
 *
 * Visit https://instantdb.com/dash to get your `appId` :)
 *
 * @example
 *  import { init } from "@instantdb/react"
 *
 *  const db = init({ appId: "my-app-id" })
 *
 *  // You can also provide a schema for type safety and editor autocomplete!
 *
 *  import { init } from "@instantdb/react"
 *  import schema from ""../instant.schema.ts";
 *
 *  const db = init({ appId: "my-app-id", schema })
 *
 *  // To learn more: https://instantdb.com/docs/modeling-data
 */
function init(
// Allows config with missing `useDateObjects`, but keeps `UseDates`
// as a non-nullable in the InstantConfig type.
config) {
    return new InstantNextDatabase_tsx_1.InstantNextDatabase(config, {
        '@instantdb/react': version_ts_1.default,
    });
}
