// @triplit/db/dist/kv-store/utils/scoped-store.js
export class ScopedKVStore {
    store;
    prefix;
    constructor(store, prefix) {
        this.store = store;
        this.prefix = prefix;
    }
    scope(scope) {
        return new ScopedKVStore(this.store, [...this.prefix, ...scope]);
    }
    get(key) {
        return this.store.get(key, this.prefix);
    }
    set(key, value) {
        return this.store.set(key, value, this.prefix);
    }
    delete(key) {
        return this.store.delete(key, this.prefix);
    }
    scan(options) {
        return this.store.scan(options, this.prefix);
    }
    scanValues(options) {
        return this.store.scanValues(options, this.prefix);
    }
    clear() {
        return this.store.clear(this.prefix);
    }
    count(options) {
        return this.store.count(options, this.prefix);
    }
    // TODO: I guess this should be scoped for correctness
    transact() {
        return new ScopedKVStoreTransaction(this.store.transact(), this.prefix);
    }
    applyEdits(sets, deletes) {
        return this.store.applyEdits(sets, deletes);
    }
}
export class ScopedKVStoreTransaction {
    tx;
    prefix;
    constructor(tx, prefix) {
        this.tx = tx;
        this.prefix = prefix;
    }
    scope(prefix) {
        return new ScopedKVStoreTransaction(this.tx, [...this.prefix, ...prefix]);
    }
    get(key) {
        return this.tx.get(key, this.prefix);
    }
    set(key, value) {
        return this.tx.set(key, value, this.prefix);
    }
    delete(key) {
        return this.tx.delete(key, this.prefix);
    }
    scan(options) {
        return this.tx.scan(options, this.prefix);
    }
    scanValues(options) {
        return this.tx.scanValues(options, this.prefix);
    }
    clear() {
        return this.tx.clear(this.prefix);
    }
    count(options) {
        return this.tx.count(options, this.prefix);
    }
    // TODO: these are a little odd on the scoped store
    commit() {
        return this.tx.commit();
    }
    cancel() {
        return this.tx.cancel();
    }
    get status() {
        return this.tx.status;
    }
}
