// @neondatabase/auth/dist/better-auth-react-adapter-DswZ_kCQ.mjs
import { t as NeonAuthAdapterCore } from "./adapter-core-D4V1J06B.mjs";
import { createAuthClient } from "better-auth/react";

//#region src/adapters/better-auth-react/better-auth-react-adapter.ts
/**
* Internal implementation class - use BetterAuthReactAdapter factory function instead
*/
var BetterAuthReactAdapterImpl = class extends NeonAuthAdapterCore {
	_betterAuth;
	constructor(betterAuthClientOptions) {
		super(betterAuthClientOptions);
		this._betterAuth = createAuthClient(this.betterAuthOptions);
	}
	getBetterAuthInstance() {
		return this._betterAuth;
	}
};
/**
* Factory function that returns an adapter builder.
* The builder is called by createClient/createAuthClient with the URL.
*
* @param options - Optional adapter configuration (baseURL is injected separately)
* @returns A builder function that creates the adapter instance
*
* @example
* ```typescript
* const client = createClient({
*   auth: {
*     url: 'https://auth.example.com',
*     adapter: BetterAuthReactAdapter(),
*   },
*   dataApi: { url: 'https://data-api.example.com' },
* });
* ```
*/
function BetterAuthReactAdapter(options) {
	return (url, fetchOptions) => new BetterAuthReactAdapterImpl({
		baseURL: url,
		...options,
		fetchOptions: {
			...options?.fetchOptions,
			headers: {
				...options?.fetchOptions?.headers,
				...fetchOptions?.headers
			}
		}
	});
}

//#endregion
export { BetterAuthReactAdapter as t };
