// @neondatabase/auth/dist/neon-auth-BEYvHA5c.mjs
import { n as BetterAuthVanillaAdapter } from "./supabase-adapter-BbSvEBy0.mjs";

//#region src/neon-auth.ts
/**
* Create a NeonAuth instance that exposes the appropriate API based on the adapter.
*
* @param url - The auth service URL (e.g., 'https://auth.example.com')
* @param config - Configuration with adapter builder
* @returns NeonAuth instance with the adapter's API exposed directly
*
* @example SupabaseAuthAdapter - Supabase-compatible API
* ```typescript
* import { createAuthClient, SupabaseAuthAdapter } from '@neondatabase/auth';
*
* const auth = createAuthClient('https://auth.example.com', {
*   adapter: SupabaseAuthAdapter(),
* });
*
* // Supabase-compatible methods
* await auth.signInWithPassword({ email, password });
* await auth.getSession();
* ```
*
* @example BetterAuthVanillaAdapter - Direct Better Auth API
* ```typescript
* import { createAuthClient, BetterAuthVanillaAdapter } from '@neondatabase/auth';
*
* const auth = createAuthClient('https://auth.example.com', {
*   adapter: BetterAuthVanillaAdapter(),
* });
*
* // Direct Better Auth API access
* await auth.signIn.email({ email, password });
* await auth.signUp.email({ email, password, name: 'John' });
* await auth.getSession();
* ```
*
* @example BetterAuthReactAdapter - Better Auth with React hooks
* ```typescript
* import { createAuthClient, BetterAuthReactAdapter } from '@neondatabase/auth';
*
* const auth = createAuthClient('https://auth.example.com', {
*   adapter: BetterAuthReactAdapter(),
* });
*
* // Direct Better Auth API with React hooks
* await auth.signIn.email({ email, password });
* const session = auth.useSession(); // React hook
* ```
*/
function createInternalNeonAuth(url, config) {
	const adapterBuilder = config?.adapter ?? BetterAuthVanillaAdapter();
	const { fetchOptions } = config ?? {};
	const adapter = adapterBuilder(url, fetchOptions);
	const allowAnonymous = config?.allowAnonymous ?? false;
	if (!(typeof adapter.initialize === "function")) return {
		getJWTToken: () => adapter.getJWTToken(allowAnonymous),
		adapter: adapter.getBetterAuthInstance()
	};
	return {
		getJWTToken: () => adapter.getJWTToken(allowAnonymous),
		adapter
	};
}
function createAuthClient(url, config) {
	return createInternalNeonAuth(url, config).adapter;
}

//#endregion
export { createInternalNeonAuth as n, createAuthClient as t };
