// @neondatabase/auth/dist/supabase-adapter-BbSvEBy0.mjs
import { i as CURRENT_TAB_CLIENT_ID, n as BETTER_AUTH_METHODS_CACHE, r as BETTER_AUTH_METHODS_HOOKS, t as NeonAuthAdapterCore } from "./adapter-core-D4V1J06B.mjs";
import { a as createAuthError, i as AuthErrorCode, l as isAuthError, n as mapBetterAuthSession, r as normalizeBetterAuthError, t as mapBetterAuthIdentity } from "./better-auth-helpers-Bkezghej.mjs";
import { createAuthClient, getGlobalBroadcastChannel } from "better-auth/client";
import "@supabase/auth-js";
import { base64url, decodeJwt, decodeProtectedHeader } from "jose";

//#region src/adapters/better-auth-vanilla/better-auth-vanilla-adapter.ts
/**
* Internal implementation class - use BetterAuthVanillaAdapter factory function instead
*/
var BetterAuthVanillaAdapterImpl = class extends NeonAuthAdapterCore {
	_betterAuth;
	constructor(betterAuthClientOptions) {
		super(betterAuthClientOptions);
		this._betterAuth = createAuthClient(this.betterAuthOptions);
	}
	getBetterAuthInstance() {
		return this._betterAuth;
	}
};
/**
* Factory function that returns an adapter builder.
* The builder is called by createClient/createAuthClient with the URL.
*
* @param options - Optional adapter configuration (baseURL is injected separately)
* @returns A builder function that creates the adapter instance
*
* @example
* ```typescript
* const client = createClient({
*   auth: {
*     url: 'https://auth.example.com',
*     adapter: BetterAuthVanillaAdapter(),
*   },
*   dataApi: { url: 'https://data-api.example.com' },
* });
* ```
*/
function BetterAuthVanillaAdapter(options) {
	return (url, fetchOptions) => new BetterAuthVanillaAdapterImpl({
		baseURL: url,
		...options,
		fetchOptions: {
			...options?.fetchOptions,
			headers: {
				...options?.fetchOptions?.headers,
				...fetchOptions?.headers
			}
		}
	});
}

//#endregion
//#region src/adapters/supabase/supabase-adapter.ts
/**
* Duck-type check for Better Auth API errors.
* Replaces `instanceof APIError` to avoid importing server-side code.
*/
function isBetterAuthAPIError(error) {
	return error !== null && typeof error === "object" && "status" in error && typeof error.status === "number";
}
/**
* Internal implementation class - use SupabaseAuthAdapter factory function instead
*/
var SupabaseAuthAdapterImpl = class extends NeonAuthAdapterCore {
	admin = void 0;
	mfa = void 0;
	oauth = void 0;
	_betterAuth;
	_stateChangeEmitters = /* @__PURE__ */ new Map();
	constructor(betterAuthClientOptions) {
		super(betterAuthClientOptions);
		this._betterAuth = createAuthClient(this.betterAuthOptions);
		/**
		* useSession() - Automatic Session Management
		*
		* Enabled by Default:
		* - ✅ Refetch on Window Focus: Automatically refetches session when user returns to the tab
		* - ✅ Cross-Tab Sync: Syncs session state across all browser tabs (sign out in one = sign out in all)
		* - ✅ Online/Offline Detection: Refetches session when network connection is restored
		* - ❌ Interval Polling: Disabled (refetchInterval: 0)
		*
		* Returns:
		* - data: Session object (user + session)
		* - isPending: Loading state
		* - isRefetching: Refetch in progress
		* - error: Error object if any
		* - refetch(): Manual refetch function
		*
		* Customize with:
		* createAuthClient({
		*   sessionOptions: {
		*     refetchOnWindowFocus: true,  // default
		*     refetchInterval: 0,           // default (seconds, 0 = off)
		*     refetchWhenOffline: false     // default
		*   }
		* })
		*/
		this._betterAuth.useSession.subscribe((value) => {
			if (!value.data?.session || !value.data?.user) {
				BETTER_AUTH_METHODS_CACHE.clearSessionCache();
				return;
			}
		});
		getGlobalBroadcastChannel().subscribe((message) => {
			if (message.clientId === CURRENT_TAB_CLIENT_ID) return;
			if (message.data && "sessionData" in message.data) {
				const sessionData = message.data.sessionData;
				const trigger = message.data.trigger;
				if (sessionData) BETTER_AUTH_METHODS_CACHE.setCachedSession(sessionData);
				else BETTER_AUTH_METHODS_CACHE.clearSessionCache();
				const supabaseSession = sessionData ? mapBetterAuthSession(sessionData.session, sessionData.user) : null;
				const promises = [...this._stateChangeEmitters.values()].map((subscription) => {
					try {
						return Promise.resolve(subscription.callback(trigger, supabaseSession));
					} catch {
						return Promise.resolve();
					}
				});
				Promise.allSettled(promises);
			}
		});
	}
	getBetterAuthInstance() {
		return this._betterAuth;
	}
	initialize = async () => {
		try {
			const session = await this.getSession();
			if (session.error) throw session.error;
			return {
				data: session.data,
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: { session: null },
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: { session: null },
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	async getSession(options) {
		try {
			const currentSession = await this._betterAuth.getSession(options?.forceFetch ? { fetchOptions: { headers: { "X-Force-Fetch": "true" } } } : void 0);
			if (!currentSession.data?.session) return {
				data: { session: null },
				error: null
			};
			return {
				data: { session: mapBetterAuthSession(currentSession.data.session, currentSession.data.user) },
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: { session: null },
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: { session: null },
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	}
	refreshSession = async () => {
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.error) throw sessionResult.error;
			return {
				data: {
					user: sessionResult.data.session?.user ?? null,
					session: sessionResult.data.session
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	setSession = async () => {
		return {
			data: {
				user: null,
				session: null
			},
			error: createAuthError(AuthErrorCode.NotImplemented, "setSession() is not supported by Better Auth. Use signInWithPassword() instead.")
		};
	};
	signUp = async (credentials) => {
		try {
			if ("email" in credentials && credentials.email && credentials.password) {
				const displayName = credentials.options?.data && "displayName" in credentials.options.data && typeof credentials.options.data.displayName === "string" ? credentials.options.data.displayName : "";
				const result = await this._betterAuth.signUp.email({
					email: credentials.email,
					password: credentials.password,
					name: displayName,
					callbackURL: credentials.options?.emailRedirectTo,
					...credentials.options?.data
				});
				if (result.error) throw normalizeBetterAuthError(result.error);
				const sessionResult = await this.getSession();
				if (!sessionResult.data.session?.user) throw createAuthError(AuthErrorCode.SessionNotFound, "Failed to retrieve user session");
				return {
					data: {
						user: sessionResult.data.session.user,
						session: sessionResult.data.session
					},
					error: null
				};
			} else if ("phone" in credentials && credentials.phone) throw createAuthError(AuthErrorCode.PhoneProviderDisabled, "Phone sign-up not supported");
			else throw createAuthError(AuthErrorCode.ValidationFailed, "Invalid credentials format");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	signInAnonymously = async () => {
		return {
			data: {
				user: null,
				session: null
			},
			error: createAuthError(AuthErrorCode.AnonymousProviderDisabled, `Anonymous sign-in is not supported. To allow unauthenticated access with an anonymous JWT, use the allowAnonymous option in your Auth configuration instead.`)
		};
	};
	signInWithPassword = async (credentials) => {
		try {
			if ("email" in credentials && credentials.email) {
				const result = await this._betterAuth.signIn.email({
					email: credentials.email,
					password: credentials.password
				});
				if (result.error) throw normalizeBetterAuthError(result.error);
				const sessionResult = await this.getSession();
				if (!sessionResult.data.session?.user) throw createAuthError(AuthErrorCode.SessionNotFound, "Failed to retrieve user session");
				return {
					data: {
						user: sessionResult.data.session.user,
						session: sessionResult.data.session
					},
					error: null
				};
			} else if ("phone" in credentials && credentials.phone) throw createAuthError(AuthErrorCode.PhoneProviderDisabled, "Phone sign-in not supported");
			else throw createAuthError(AuthErrorCode.ValidationFailed, "Invalid credentials format");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	signInWithOAuth = async (credentials) => {
		try {
			const { provider, options } = credentials;
			await this._betterAuth.signIn.social({
				provider,
				scopes: options?.scopes?.split(" "),
				disableRedirect: options?.skipBrowserRedirect,
				callbackURL: options?.redirectTo || (globalThis.window === void 0 ? "" : globalThis.location.origin)
			});
			return {
				data: {
					provider,
					url: options?.redirectTo || (globalThis.window === void 0 ? "" : globalThis.location.origin)
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					provider: credentials.provider,
					url: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					provider: credentials.provider,
					url: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	signInWithOtp = async (credentials) => {
		try {
			if ("phone" in credentials) return {
				data: {
					user: null,
					session: null,
					messageId: void 0
				},
				error: createAuthError(AuthErrorCode.PhoneProviderDisabled, "Phone OTP authentication is not supported. Use email-based authentication instead.")
			};
			if ("email" in credentials) {
				await this._betterAuth.emailOtp.sendVerificationOtp({
					email: credentials.email,
					type: "sign-in"
				});
				return {
					data: {
						user: null,
						session: null,
						messageId: void 0
					},
					error: null
				};
			}
			throw createAuthError(AuthErrorCode.ValidationFailed, "Invalid OTP credentials format");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null,
					messageId: void 0
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null,
					messageId: void 0
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	signInWithIdToken = async (credentials) => {
		try {
			const result = await this._betterAuth.signIn.social({
				provider: credentials.provider,
				idToken: {
					token: credentials.token,
					accessToken: credentials.access_token,
					nonce: credentials.nonce
				}
			});
			if (result.error) throw normalizeBetterAuthError(result.error);
			if (!("user" in result.data) || !result.data.user) throw createAuthError(AuthErrorCode.OAuthCallbackFailed, "Failed to sign in with ID token");
			const session = await this.getSession();
			if (session.error || !session.data.session) throw session.error || createAuthError(AuthErrorCode.SessionNotFound, "Failed to get session");
			return {
				data: {
					user: session.data.session.user,
					session: session.data.session
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	signInWithSSO = async (params) => {
		const attemptedWith = "providerId" in params ? `provider ID: ${params.providerId}` : `domain: ${"domain" in params ? params.domain : "unknown"}`;
		return {
			data: null,
			error: createAuthError(AuthErrorCode.SsoProviderDisabled, `Better Auth does not support enterprise SAML SSO. Attempted with ${attemptedWith}. Use signInWithOAuth() for OAuth providers instead.`)
		};
	};
	signInWithWeb3 = async (credentials) => {
		const attemptedChain = credentials.chain;
		return {
			data: {
				user: null,
				session: null
			},
			error: createAuthError(AuthErrorCode.Web3ProviderDisabled, `Better Auth does not support Web3 authentication. Attempted with chain: ${attemptedChain}. Supported: OAuth, email/password, magic link.`)
		};
	};
	signOut = async () => {
		try {
			const result = await this._betterAuth.signOut();
			if (result.error) throw normalizeBetterAuthError(result.error);
			return { error: null };
		} catch (error) {
			if (isAuthError(error)) return { error };
			if (isBetterAuthAPIError(error)) return { error: normalizeBetterAuthError(error) };
			throw error;
		}
	};
	getUser = async () => {
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.error || !sessionResult.data.session) throw sessionResult.error || createAuthError(AuthErrorCode.SessionNotFound, "No user session found");
			return {
				data: { user: sessionResult.data.session.user },
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: { user: null },
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: { user: null },
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	getClaims = async (jwtArg) => {
		try {
			let jwt = jwtArg;
			if (!jwt) {
				const sessionResult = await this.getSession();
				if (sessionResult.error || !sessionResult.data?.session) throw sessionResult.error || createAuthError(AuthErrorCode.SessionNotFound, "No user session found");
				jwt = sessionResult.data.session.access_token;
			}
			if (!jwt) throw createAuthError(AuthErrorCode.SessionNotFound, "No access token found");
			if (jwt.split(".").length !== 3) throw createAuthError(AuthErrorCode.BadJwt, "Invalid token format");
			return {
				data: {
					header: decodeProtectedHeader(jwt),
					claims: decodeJwt(jwt),
					signature: base64url.decode(jwt.split(".")[2])
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: null,
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: null,
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	updateUser = async (attributes) => {
		try {
			if (attributes.password) throw createAuthError(AuthErrorCode.FeatureNotSupported, "The password cannot be updated through the updateUser method, use the changePassword method instead.");
			if (attributes.email) throw createAuthError(AuthErrorCode.FeatureNotSupported, "The email cannot be updated through the updateUser method, use the changeEmail method instead.");
			const result = await this._betterAuth.updateUser({ ...attributes.data });
			if (result.data?.status) throw createAuthError(AuthErrorCode.InternalError, "Failed to update user");
			if (result?.error) throw normalizeBetterAuthError(result.error);
			const updatedSessionResult = await this.getSession({ forceFetch: true });
			if (!updatedSessionResult.data.session) throw createAuthError(AuthErrorCode.SessionNotFound, "Failed to retrieve updated user");
			return {
				data: { user: updatedSessionResult.data.session.user },
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: { user: null },
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: { user: null },
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	getUserIdentities = async () => {
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.error || !sessionResult.data.session) throw sessionResult.error || createAuthError(AuthErrorCode.SessionNotFound, "No user session found");
			const result = await this._betterAuth.listAccounts();
			if (!result) throw createAuthError(AuthErrorCode.InternalError, "Failed to list accounts");
			if (result.error) throw normalizeBetterAuthError(result.error);
			const identitiesPromises = result.data.map(async (account) => {
				let accountInfo = null;
				try {
					accountInfo = (await this._betterAuth.accountInfo({ query: { accountId: account.accountId } })).data;
				} catch (error) {
					console.warn(`Failed to get account info for ${account.providerId}:`, error);
				}
				return mapBetterAuthIdentity(account, accountInfo ?? null);
			});
			return {
				data: { identities: await Promise.all(identitiesPromises) },
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: null,
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: null,
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	linkIdentity = async (credentials) => {
		const provider = credentials.provider;
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.error || !sessionResult.data.session) throw sessionResult.error || createAuthError(AuthErrorCode.SessionNotFound, "No user session found");
			if ("token" in credentials) {
				const result$1 = await this._betterAuth.linkSocial({
					provider,
					idToken: {
						token: credentials.token,
						accessToken: credentials.access_token,
						nonce: credentials.nonce
					}
				});
				if (result$1.error) throw normalizeBetterAuthError(result$1.error);
				return {
					data: {
						user: sessionResult.data.session.user,
						session: sessionResult.data.session,
						provider,
						url: result$1.data?.url
					},
					error: null
				};
			}
			const callbackURL = credentials.options?.redirectTo || (globalThis.window === void 0 ? "" : globalThis.location.origin);
			const scopes = credentials.options?.scopes?.split(" ").filter((s) => s.length > 0);
			const result = await this._betterAuth.linkSocial({
				provider,
				callbackURL,
				errorCallbackURL: callbackURL ? `${callbackURL}?error=linking-failed` : void 0,
				scopes
			});
			if (result.error) throw normalizeBetterAuthError(result.error);
			return {
				data: {
					provider,
					url: result.data?.url,
					user: sessionResult.data.session.user,
					session: sessionResult.data.session
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					provider,
					url: null,
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					provider,
					url: null,
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	unlinkIdentity = async (identity) => {
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.error || !sessionResult.data.session) throw sessionResult.error || createAuthError(AuthErrorCode.SessionNotFound, "No user session found");
			const identities = await this.getUserIdentities();
			if (identities.error || !identities.data) throw identities.error || createAuthError(AuthErrorCode.InternalError, "Failed to fetch identities");
			const targetIdentity = identities.data.identities.find((i) => i.id === identity.identity_id);
			if (!targetIdentity) throw createAuthError(AuthErrorCode.IdentityNotFound, "Identity not found");
			const providerId = targetIdentity.provider;
			const accountId = targetIdentity.identity_id;
			const result = await this._betterAuth.unlinkAccount({
				providerId,
				accountId
			});
			if (result?.error) throw normalizeBetterAuthError(result.error);
			const updatedSession = await this.getSession({ forceFetch: true });
			if (updatedSession.data.session) BETTER_AUTH_METHODS_HOOKS["updateUser"].onSuccess(updatedSession.data.session);
			return {
				data: {},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: null,
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: null,
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	verifyOtp = async (params) => {
		try {
			if ("phone" in params && params.phone) return {
				data: {
					user: null,
					session: null
				},
				error: createAuthError(AuthErrorCode.PhoneProviderDisabled, "Phone OTP verification is not supported. Use email-based authentication instead.")
			};
			if ("email" in params && params.type === "magiclink") return {
				data: {
					user: null,
					session: null
				},
				error: createAuthError(AuthErrorCode.MagicLinkNotSupported, "Magic link verification is not supported. Use email OTP authentication instead.")
			};
			if ("email" in params && params.email) return await this.verifyEmailOtp(params);
			if ("token_hash" in params && params.token_hash) throw createAuthError(AuthErrorCode.FeatureNotSupported, "Token hash verification not supported");
			throw createAuthError(AuthErrorCode.ValidationFailed, "Invalid OTP verification parameters");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	resetPasswordForEmail = async (email, options) => {
		try {
			const result = await this._betterAuth.requestPasswordReset({
				email,
				redirectTo: options?.redirectTo || (globalThis.window === void 0 ? "" : globalThis.location.origin)
			});
			if (result?.error) throw normalizeBetterAuthError(result.error);
			return {
				data: {},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: null,
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: null,
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	reauthenticate = async () => {
		try {
			const newSession = await this.getSession();
			if (newSession.error || !newSession.data.session) throw newSession.error || createAuthError(AuthErrorCode.SessionNotFound, "No session found");
			return {
				data: {
					user: newSession.data.session?.user || null,
					session: newSession.data.session
				},
				error: null
			};
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	resend = async (credentials) => {
		try {
			if ("email" in credentials) {
				const { email, type, options } = credentials;
				if (type === "signup" || type === "email_change") {
					const result = await this._betterAuth.sendVerificationEmail({
						email,
						callbackURL: options?.emailRedirectTo || (globalThis.window === void 0 ? "" : globalThis.location.origin)
					});
					if (result?.error) throw normalizeBetterAuthError(result.error);
					return {
						data: {
							user: null,
							session: null
						},
						error: null
					};
				}
				throw createAuthError(AuthErrorCode.ValidationFailed, `Unsupported resend type: ${type}`);
			}
			if ("phone" in credentials) return {
				data: {
					user: null,
					session: null
				},
				error: createAuthError(AuthErrorCode.PhoneProviderDisabled, "SMS resend is not supported. Use email-based authentication instead.")
			};
			throw createAuthError(AuthErrorCode.ValidationFailed, "Invalid credentials format");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					user: null,
					session: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	exchangeCodeForSession = async (_authCode) => {
		try {
			const sessionResult = await this.getSession();
			if (sessionResult.data.session) return {
				data: {
					session: sessionResult.data.session,
					user: sessionResult.data.session.user
				},
				error: null
			};
			throw createAuthError(AuthErrorCode.OAuthCallbackFailed, "OAuth callback completed but no session was created. Make sure the OAuth callback has been processed.");
		} catch (error) {
			if (isAuthError(error)) return {
				data: {
					session: null,
					user: null
				},
				error
			};
			if (isBetterAuthAPIError(error)) return {
				data: {
					session: null,
					user: null
				},
				error: normalizeBetterAuthError(error)
			};
			throw error;
		}
	};
	onAuthStateChange = (callback) => {
		const id = crypto.randomUUID();
		const subscription = {
			id,
			callback,
			unsubscribe: () => {
				this._stateChangeEmitters.delete(id);
			}
		};
		this._stateChangeEmitters.set(id, subscription);
		this.emitInitialSession(callback);
		return { data: { subscription: {
			id,
			callback,
			unsubscribe: subscription.unsubscribe
		} } };
	};
	isThrowOnErrorEnabled = () => false;
	startAutoRefresh = async () => {};
	stopAutoRefresh = async () => {};
	async verifyEmailOtp(params) {
		const { type } = params;
		if (type === "email") {
			const result = await this._betterAuth.signIn.emailOtp({
				email: params.email,
				otp: params.token
			});
			if (result.error) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(result.error)
			};
			const sessionResult = await this.getSession({ forceFetch: true });
			if (!sessionResult.data.session) return {
				data: {
					user: null,
					session: null
				},
				error: createAuthError(AuthErrorCode.SessionNotFound, "Failed to retrieve session after OTP verification. Make sure the magic link callback has been processed.")
			};
			return {
				data: {
					user: sessionResult.data.session.user,
					session: sessionResult.data.session
				},
				error: null
			};
		}
		if (type === "magiclink") return {
			data: {
				user: null,
				session: null
			},
			error: createAuthError(AuthErrorCode.MagicLinkNotSupported, "Magic link verification is not supported. Use email OTP authentication instead.")
		};
		if (type === "signup" || type === "invite") {
			const result = await this._betterAuth.emailOtp.verifyEmail({
				email: params.email,
				otp: params.token
			});
			if (result?.error) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(result.error)
			};
			const sessionResult = await this.getSession({ forceFetch: true });
			return {
				data: {
					user: sessionResult.data.session?.user ?? null,
					session: sessionResult.data.session
				},
				error: null
			};
		}
		if (type === "recovery") {
			const checkResult = await this._betterAuth.emailOtp.checkVerificationOtp({
				email: params.email,
				otp: params.token,
				type: "forget-password"
			});
			if (checkResult.error || !checkResult.data?.success) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(checkResult.error)
			};
			return {
				data: {
					user: null,
					session: null
				},
				error: null
			};
		}
		if (type === "email_change") {
			const result = await this._betterAuth.verifyEmail({ query: {
				token: params.token,
				callbackURL: params.options?.redirectTo
			} });
			if (result?.error) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(result.error)
			};
			const sessionResult = await this.getSession({ forceFetch: true });
			if (sessionResult.error || !sessionResult.data) return {
				data: {
					user: null,
					session: null
				},
				error: sessionResult.error || createAuthError(AuthErrorCode.InternalError, "Failed to get session")
			};
			if (sessionResult.data.session) BETTER_AUTH_METHODS_HOOKS["updateUser"].onSuccess(sessionResult.data.session);
			return {
				data: {
					user: sessionResult.data?.session?.user || null,
					session: sessionResult.data?.session || null
				},
				error: null
			};
		}
		if (type === "invite") {
			const result = await this._betterAuth.organization.acceptInvitation({ invitationId: params.token });
			if (result.error) return {
				data: {
					user: null,
					session: null
				},
				error: normalizeBetterAuthError(result.error)
			};
			const sessionResult = await this.getSession({ forceFetch: true });
			if (sessionResult.error || !sessionResult.data) return {
				data: {
					user: null,
					session: null
				},
				error: sessionResult.error || createAuthError(AuthErrorCode.InternalError, "Failed to get session")
			};
			return {
				data: {
					user: sessionResult.data?.session?.user || null,
					session: sessionResult.data?.session
				},
				error: null
			};
		}
		return {
			data: {
				user: null,
				session: null
			},
			error: createAuthError(AuthErrorCode.ValidationFailed, `Unsupported email OTP type: ${type}`)
		};
	}
	async emitInitialSession(callback) {
		try {
			const { data, error } = await this.getSession();
			if (error) {
				await callback("INITIAL_SESSION", null);
				return;
			}
			await callback("INITIAL_SESSION", data.session);
		} catch {
			await callback("INITIAL_SESSION", null);
		}
	}
};
/**
* Factory function that returns an adapter builder.
* The builder is called by createClient/createAuthClient with the URL.
*
* @param options - Optional adapter configuration (baseURL is injected separately)
* @returns A builder function that creates the adapter instance
*
* @example
* ```typescript
* const client = createClient({
*   auth: {
*     url: 'https://auth.example.com',
*     adapter: SupabaseAuthAdapter(),
*   },
*   dataApi: { url: 'https://data-api.example.com' },
* });
* ```
*/
function SupabaseAuthAdapter(options) {
	return (url, fetchOptions) => new SupabaseAuthAdapterImpl({
		baseURL: url,
		...options,
		fetchOptions: {
			...options?.fetchOptions,
			headers: {
				...options?.fetchOptions?.headers,
				...fetchOptions?.headers
			}
		}
	});
}

//#endregion
export { BetterAuthVanillaAdapter as n, SupabaseAuthAdapter as t };
