// @instantdb/react-common/dist/commonjs/InstantReactRoom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstantReactRoom = exports.rooms = exports.defaultActivityStopTimeout = void 0;
exports.useTopicEffect = useTopicEffect;
exports.usePublishTopic = usePublishTopic;
exports.usePresence = usePresence;
exports.useSyncPresence = useSyncPresence;
exports.useTypingIndicator = useTypingIndicator;
const react_1 = require("react");
const useTimeout_ts_1 = require("./useTimeout.js");
exports.defaultActivityStopTimeout = 1_000;
// ------
// Topics
/**
 * Listen for broadcasted events given a room and topic.
 *
 * @see https://instantdb.com/docs/presence-and-topics
 * @example
 *  function App({ roomId }) {
 *    const room = db.room('chats', roomId);
 *    db.rooms.useTopicEffect(room, 'emoji', (message, peer) => {
 *      console.log(peer.name, 'sent', message);
 *    });
 *    // ...
 *  }
 */
function useTopicEffect(room, topic, onEvent) {
    const onEventRef = (0, react_1.useRef)(onEvent);
    onEventRef.current = onEvent;
    (0, react_1.useEffect)(() => {
        const unsub = room.core._reactor.subscribeTopic(room.type, room.id, topic, (event, peer) => {
            onEventRef.current(event, peer);
        });
        return unsub;
    }, [room.id, topic]);
}
/**
 * Broadcast an event to a room.
 *
 * @see https://instantdb.com/docs/presence-and-topics
 * @example
 * function App({ roomId }) {
 *   const room = db.room('chat', roomId);
 *   const publishTopic = db.rooms.usePublishTopic(room, "emoji");
 *
 *   return (
 *     <button onClick={() => publishTopic({ emoji: "🔥" })}>Send emoji</button>
 *   );
 * }
 *
 */
function usePublishTopic(room, topic) {
    (0, react_1.useEffect)(() => room.core._reactor.joinRoom(room.type, room.id), [room.id]);
    const publishTopic = (0, react_1.useCallback)((data) => {
        room.core._reactor.publishTopic({
            roomType: room.type,
            roomId: room.id,
            topic,
            data,
        });
    }, [room.id, topic]);
    return publishTopic;
}
// ---------
// Presence
/**
 * Listen for peer's presence data in a room, and publish the current user's presence.
 *
 * @see https://instantdb.com/docs/presence-and-topics
 * @example
 *  function App({ roomId }) {
 *    const {
 *      peers,
 *      publishPresence
 *    } = db.room(roomType, roomId).usePresence({ keys: ["name", "avatar"] });
 *
 *    // ...
 *  }
 */
function usePresence(room, opts = {}) {
    const [state, setState] = (0, react_1.useState)(() => {
        return (room.core._reactor.getPresence(room.type, room.id, opts) ?? {
            peers: {},
            isLoading: true,
        });
    });
    (0, react_1.useEffect)(() => {
        const unsub = room.core._reactor.subscribePresence(room.type, room.id, opts, (data) => {
            setState(data);
        });
        return unsub;
    }, [room.id, opts.user, opts.peers?.join(), opts.keys?.join()]);
    const publishPresence = (0, react_1.useCallback)((data) => {
        room.core._reactor.publishPresence(room.type, room.id, data);
    }, [room.type, room.id]);
    const ret = (0, react_1.useMemo)(() => {
        return {
            ...state,
            publishPresence,
        };
    }, [state, publishPresence]);
    return ret;
}
/**
 * Publishes presence data to a room
 *
 * @see https://instantdb.com/docs/presence-and-topics
 * @example
 *  function App({ roomId, nickname }) {
 *    const room = db.room('chat', roomId);
 *    db.rooms.useSyncPresence(room, { nickname });
 *  }
 */
function useSyncPresence(room, data, deps) {
    (0, react_1.useEffect)(() => room.core._reactor.joinRoom(room.type, room.id, data), [room.id]);
    (0, react_1.useEffect)(() => {
        return room.core._reactor.publishPresence(room.type, room.id, data);
    }, [room.type, room.id, deps ?? JSON.stringify(data)]);
}
// -----------------
// Typing Indicator
/**
 * Manage typing indicator state
 *
 * @see https://instantdb.com/docs/presence-and-topics
 * @example
 *  function App({ roomId }) {
 *    const room = db.room('chat', roomId);
 *    const {
 *      active,
 *      setActive,
 *      inputProps,
 *    } = db.rooms.useTypingIndicator(room, "chat-input");
 *
 *    return <input {...inputProps} />;
 *  }
 */
function useTypingIndicator(room, inputName, opts = {}) {
    const timeout = (0, useTimeout_ts_1.useTimeout)();
    const observedPresence = exports.rooms.usePresence(room, {
        keys: [inputName],
    });
    const active = (0, react_1.useMemo)(() => {
        const presenceSnapshot = room._core._reactor.getPresence(room.type, room.id);
        return opts?.writeOnly
            ? []
            : Object.values(presenceSnapshot?.peers ?? {}).filter((p) => p[inputName] === true);
    }, [opts?.writeOnly, observedPresence]);
    const setActive = (0, react_1.useCallback)((isActive) => {
        room.core._reactor.publishPresence(room.type, room.id, {
            [inputName]: isActive,
        });
        if (!isActive)
            return;
        if (opts?.timeout === null || opts?.timeout === 0)
            return;
        timeout.set(opts?.timeout ?? exports.defaultActivityStopTimeout, () => {
            room.core._reactor.publishPresence(room.type, room.id, {
                [inputName]: null,
            });
        });
    }, [room.type, room.id, inputName, opts?.timeout, timeout]);
    const onKeyDown = (0, react_1.useCallback)((e) => {
        const isEnter = opts?.stopOnEnter && e.key === 'Enter';
        const isActive = !isEnter;
        setActive(isActive);
    }, [opts.stopOnEnter, setActive]);
    const onBlur = (0, react_1.useCallback)(() => {
        setActive(false);
    }, [setActive]);
    const inputProps = (0, react_1.useMemo)(() => {
        return { onKeyDown, onBlur };
    }, [onKeyDown, onBlur]);
    return {
        active,
        setActive,
        inputProps,
    };
}
// --------------
// Hooks
exports.rooms = {
    useTopicEffect,
    usePublishTopic,
    usePresence,
    useSyncPresence,
    useTypingIndicator,
};
// ------------
// Class
class InstantReactRoom {
    core;
    /** @deprecated use `core` instead */
    _core;
    type;
    id;
    constructor(core, type, id) {
        this.core = core;
        this._core = this.core;
        this.type = type;
        this.id = id;
    }
    /**
     * @deprecated
     * `db.room(...).useTopicEffect` is deprecated. You can replace it with `db.rooms.useTopicEffect`.
     *
     * @example
     *
     * // Before
     * const room = db.room('chat', 'room-id');
     * room.useTopicEffect('emoji', (message, peer) => {  });
     *
     * // After
     * const room = db.room('chat', 'room-id');
     * db.rooms.useTopicEffect(room, 'emoji', (message, peer) => {  });
     */
    useTopicEffect = (topic, onEvent) => {
        exports.rooms.useTopicEffect(this, topic, onEvent);
    };
    /**
     * @deprecated
     * `db.room(...).usePublishTopic` is deprecated. You can replace it with `db.rooms.usePublishTopic`.
     *
     * @example
     *
     * // Before
     * const room = db.room('chat', 'room-id');
     * const publish = room.usePublishTopic('emoji');
     *
     * // After
     * const room = db.room('chat', 'room-id');
     * const publish = db.rooms.usePublishTopic(room, 'emoji');
     */
    usePublishTopic = (topic) => {
        return exports.rooms.usePublishTopic(this, topic);
    };
    /**
     * @deprecated
     * `db.room(...).usePresence` is deprecated. You can replace it with `db.rooms.usePresence`.
     *
     * @example
     *
     * // Before
     * const room = db.room('chat', 'room-id');
     * const { peers } = room.usePresence({ keys: ["name", "avatar"] });
     *
     * // After
     * const room = db.room('chat', 'room-id');
     * const { peers } = db.rooms.usePresence(room, { keys: ["name", "avatar"] });
     */
    usePresence = (opts = {}) => {
        return exports.rooms.usePresence(this, opts);
    };
    /**
     * @deprecated
     * `db.room(...).useSyncPresence` is deprecated. You can replace it with `db.rooms.useSyncPresence`.
     *
     * @example
     *
     * // Before
     * const room = db.room('chat', 'room-id');
     * room.useSyncPresence(room, { nickname });
     *
     * // After
     * const room = db.room('chat', 'room-id');
     * db.rooms.useSyncPresence(room, { nickname });
     */
    useSyncPresence = (data, deps) => {
        return exports.rooms.useSyncPresence(this, data, deps);
    };
    /**
     * @deprecated
     * `db.room(...).useTypingIndicator` is deprecated. You can replace it with `db.rooms.useTypingIndicator`.
     *
     * @example
     *
     * // Before
     * const room = db.room('chat', 'room-id');
     * const typing = room.useTypingIndiactor(room, 'chat-input');
     *
     * // After
     * const room = db.room('chat', 'room-id');
     * const typing = db.rooms.useTypingIndiactor(room, 'chat-input');
     */
    useTypingIndicator = (inputName, opts = {}) => {
        return exports.rooms.useTypingIndicator(this, inputName, opts);
    };
}
exports.InstantReactRoom = InstantReactRoom;
