// @instantdb/react/dist/esm/next-ssr/index.js
import version from "../version.js";
import { InstantNextDatabase } from "./InstantNextDatabase.js";
export { getUnverifiedUserFromInstantCookie, getUserFromInstantCookie, } from "./getUserFromInstantCookie.js";
export { InstantNextDatabase } from "./InstantNextDatabase.js";
export { InstantSuspenseProvider } from "./InstantSuspenseProvider.js";
export { createInstantRouteHandler } from '@instantdb/core';
/**
 *
 * The first step: init your application!
 *
 * Visit https://instantdb.com/dash to get your `appId` :)
 *
 * @example
 *  import { init } from "@instantdb/react"
 *
 *  const db = init({ appId: "my-app-id" })
 *
 *  // You can also provide a schema for type safety and editor autocomplete!
 *
 *  import { init } from "@instantdb/react"
 *  import schema from ""../instant.schema.ts";
 *
 *  const db = init({ appId: "my-app-id", schema })
 *
 *  // To learn more: https://instantdb.com/docs/modeling-data
 */
export function init(
// Allows config with missing `useDateObjects`, but keeps `UseDates`
// as a non-nullable in the InstantConfig type.
config) {
    return new InstantNextDatabase(config, {
        '@instantdb/react': version,
    });
}
