// @instantdb/react/dist/commonjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstantRouteHandler = exports.StoreInterface = exports.setInstantWarningsEnabled = exports.SyncTableCallbackEventType = exports.InstantReactAbstractDatabase = exports.InstantAPIError = exports.i = exports.Cursors = exports.InstantReactWebDatabase = exports.init_experimental = exports.init = exports.lookup = exports.tx = exports.id = void 0;
const core_1 = require("@instantdb/core");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return core_1.id; } });
Object.defineProperty(exports, "tx", { enumerable: true, get: function () { return core_1.tx; } });
Object.defineProperty(exports, "lookup", { enumerable: true, get: function () { return core_1.lookup; } });
Object.defineProperty(exports, "i", { enumerable: true, get: function () { return core_1.i; } });
Object.defineProperty(exports, "InstantAPIError", { enumerable: true, get: function () { return 
    // error
    core_1.InstantAPIError; } });
Object.defineProperty(exports, "SyncTableCallbackEventType", { enumerable: true, get: function () { return 
    // sync table enums
    core_1.SyncTableCallbackEventType; } });
Object.defineProperty(exports, "setInstantWarningsEnabled", { enumerable: true, get: function () { return core_1.setInstantWarningsEnabled; } });
Object.defineProperty(exports, "StoreInterface", { enumerable: true, get: function () { return core_1.StoreInterface; } });
Object.defineProperty(exports, "createInstantRouteHandler", { enumerable: true, get: function () { return core_1.createInstantRouteHandler; } });
const react_common_1 = require("@instantdb/react-common");
Object.defineProperty(exports, "InstantReactAbstractDatabase", { enumerable: true, get: function () { return react_common_1.InstantReactAbstractDatabase; } });
const InstantReactWebDatabase_ts_1 = __importDefault(require("./InstantReactWebDatabase.js"));
exports.InstantReactWebDatabase = InstantReactWebDatabase_ts_1.default;
const init_ts_1 = require("./init.js");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return init_ts_1.init; } });
Object.defineProperty(exports, "init_experimental", { enumerable: true, get: function () { return init_ts_1.init_experimental; } });
const Cursors_tsx_1 = require("./Cursors.js");
Object.defineProperty(exports, "Cursors", { enumerable: true, get: function () { return Cursors_tsx_1.Cursors; } });
